### PR TITLE
Finish porting advapi32 to v0.3

### PR DESCRIPTION
**Pull request by austinwagner**
Used @Susurrus's fork as a base.

All functions exported by advapi32 that can be found in the SDK headers* have been added to their corresponding Rust files with the exception of SetServiceBits (um/LMServer.h). If a Rust version of a header already existed then only the required pieces were added to the file. If the Rust file didn't exist already the full header was ported to the extent I was able to.

*139 functions are exported from the DLL but have no declaration in any header in the latest Windows SDK.

### CODE REVIEW

**Comment by austinwagner** [comment]
Unused macros? Is this a CI issue or did I break something? I tried building each of the features that I added or modified with Rust stable and I don't get any errors.

**Comment by retep998** [comment]
It's a new nightly lint that got caught in `deny(unused)`.

**Comment by Susurrus** [comment]
@austinwagner That issue is being addressed in #446.

**Comment by Susurrus** [comment]
@austinwagner Your header import declarations in `build.rs` aren't listed alphabetically, which is why they're failing the tests.

**Comment by austinwagner** [comment]
Yes it's clear why it's failing this time. 🙂 Apparently I'm not good at quickly eyeballing alphabetical order. I'm going to fix it as soon as I get home from work. The error message shows me all of the build.rs sorting problems, but is there a way to quickly sort the imports in the other files? I haven't tried rustfmt because I don't know if it will keep the rest of the file formatted the way that our Benevolent Bunny likes.

**Comment by Susurrus** [comment]
I believe the only errors are in `build.rs`, so I don't think you need to reorder anything in any of the actual `.rs` files. But regarding `rustfmt`, #444 is tracking its suitability for `winapi-rs`.

**Comment by austinwagner** [comment]
Is using the `memset` and `strnlen` from `ntdll` acceptable? I know that anything exported by `ntdll` is considered to be an internal function, but I doubt they're going away. If it's a problem I can replace them with simple loop implementations though.

**Comment by retep998** [comment]
`core::ptr::write_bytes` is a thing you know.

**Comment by austinwagner** [comment]
No I did not know. My search for intrinsics led me to a page that said they will likely never make it to stable. "...they should be used through stabilized interfaces in the rest of the standard library" makes it sound like you can't access the intrinsics on stable when `no_std` is set.

**Comment by retep998** [comment]
`core::ptr::write_bytes` is not an intrinsic, it is a stable wrapper around an intrinsic.

**Comment by Susurrus** [comment]
@austinwagner Needs rebasing now that #448 was merged.

**Comment by austinwagner** [comment]
@Susurrus Is rebasing the right approach or should I just merge again like I did for smaller conflicts? I've never contributed to a large, popular repo so I don't know what's considered best practice.

**Comment by Susurrus** [comment]
I always rebase rather than merge as if bisecting ever needs to be done, it's easier with less merge nodes. I also find it easier to review PRs that don't have a ton of extraneous "merge ..." nodes.

**Comment by Susurrus** [comment]
I also rebase and squash/edit commits to get rid of all of those small "fix ..." commits (like you have in this commit). The way I look at my commits is that they should always result in a compilable source tree. That's not always possible, but I strive for it. So merging smaller fix commits into the original commit that broke thing before merging is nice.

**Comment by austinwagner** [comment]
You answered my next question before I got to ask it 😃 I'll take care of this as soon as I'm able so the PR can be nice and clean!

I do have one more question for now: was splitting up the commits by header a good approach? I specifically used hunk staging to isolate each one even though I didn't actually start making commits until I finished the porting.

**Comment by retep998** [comment]
You have some commits which are actually rebased versions of commits by @Susurrus which makes it a lot harder to review this PR. Once those commits are rebased away so this PR is just your commits, I'll be able to review it.

**Comment by Susurrus** [comment]
@austinwagner Re the single-header commits, I haven't seen @retep998 complain about it yet, but it at least makes it easy for me to review and check that I have all the necessary changes for a single header/family of headers to be enabled.

**Comment by austinwagner** [comment]
All cleaned up with a rebase

**Comment by Susurrus** [comment]
timezoneapi's functions are in `kernel32`, so that needs to specified as a dependency. Additionally, those functions are all in `lib/kernel32/src/lib.rs`. Can you remove those functions from that file in that same commit?

File: src/shared/evntrace.rs [hunk 1]
```diff
@@ -0,0 +1,961 @@
...
+    EventsLost: ULONG,
+    BuffersWritten: ULONG,
+    LogBuffersLost: ULONG,
+    RealTimeBuffersLost: ULONG,
+    LoggerThreadId: HANDLE,
+    LogFileNameOffset: ULONG,
+    LoggerNameOffset: ULONG,
+}}
+pub type PEVENT_TRACE_PROPERTIES = *mut EVENT_TRACE_PROPERTIES;
+// TODO: The single item structs are defined with the ":" bitfield syntax
```
Comment (Susurrus) [inline src/shared/evntrace.rs]: Does this TODO still need to be resolved?

File: src/shared/evntprov.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+pub unsafe fn EventDataDescCreate(
+    EventDataDescriptor: PEVENT_DATA_DESCRIPTOR, DataPtr: *const VOID, DataSize: ULONG
+) {
+    (*EventDataDescriptor).Ptr = DataPtr as ULONGLONG;
+    (*EventDataDescriptor).Size = DataSize;
+    *(*EventDataDescriptor).u.Reserved_mut() = 0;
+}
+#[inline(always)]
+pub unsafe fn EventDescCreate(
+    EventDescriptor: PEVENT_DESCRIPTOR, Id: USHORT, Version: UCHAR, Channel: UCHAR, Level: UCHAR,
```
Comment (Susurrus) [inline src/shared/evntprov.rs]: These arguments all need to be on different lines.

Comment (Susurrus) [reply src/shared/evntprov.rs]: This applies to the following function declarations as well.

File: src/shared/evntrace.rs (hunk 1 above)
Comment (austinwagner) [reply src/shared/evntrace.rs]: Oh it looks like I simply forgot to remove the comment after I found the macro to handle bitfields.

File: src/shared/evntprov.rs (hunk 2 above)
Comment (austinwagner) [reply src/shared/evntprov.rs]: I didn't realize that style applied to function definitions. Thought it was only for the declarations. I'll go through all the files and validate that I don't have the same issue elsewhere.

Comment (austinwagner) [reply src/shared/evntprov.rs]: @retep998 Looking at some existing `#[inline]` functions, if everything up to and including the open curly bracket fits within the 99 character limit then the parameters should all stay on a single line, correct?

Comment (retep998) [reply src/shared/evntprov.rs]: For function *definitions* (functions with bodies), if the function signature fits within a single line then it can stay on one line but it's not required. If it isn't a single line, then it should follow the same one parameter per line that all other definitions follow.

However, you shouldn't be using `#[inline(always)]`. Function definitions should always use `#[inline]`.

File: src/shared/sddl.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        StringSecurityDescriptorLen: PULONG,
+    ) -> BOOL;
+    pub fn ConvertSecurityDescriptorToStringSecurityDescriptorW(
+        SecurityDescriptor: PSECURITY_DESCRIPTOR,
+        RequestedStringSDRevision: DWORD,
+        SecurityInformation: SECURITY_INFORMATION,
+        StringSecurityDescriptor: *mut LPWSTR,
+        StringSecurityDescriptorLen: PULONG,
+    ) -> BOOL;
+}
```
Comment (retep998) [inline src/shared/sddl.rs]: Don't forget end of file newlines.

File: src/shared/wmistr.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    WnodeHeader: WNODE_HEADER,
+    SizeNeeded: ULONG,
+}}
+pub type PWNODE_TOO_SMALL = *mut WNODE_TOO_SMALL;
+#[cfg(target_arch = "x86")]
+const __SIZE_OF_WMIREGGUIDW_u: usize = 4;
+#[cfg(target_arch = "x86_64")]
+const __SIZE_OF_WMIREGGUIDW_u: usize = 8;
+UNION2!{union WMIREGGUIDW_u {
+    [u8; __SIZE_OF_WMIREGGUIDW_u],
```
Comment (retep998) [inline src/shared/wmistr.rs]: I recently added support for `UNION2` to have two storages specified, one for 32bit and one for 64bit. Also, an array of `u8` is usually the wrong thing because the union needs to have the correct alignment (otherwise when placed inside a struct it could have the wrong position resulting in issues like https://github.com/retep998/winapi-rs/issues/452).

Comment (austinwagner) [reply src/shared/wmistr.rs]: How do you determine which to use then? I just got the sizes by looking at the output of `sizeof()` in C++ for both 32 and 64 bit builds.

Comment (retep998) [reply src/shared/wmistr.rs]:
```C++
char const * type_for_alignment(uintptr_t align) {
    switch (align) {
    case 1: return "u8";
    case 2: return "u16";
    case 4: return "u32";
    case 8: return "u64";
    }
    throw;
}
#define PRINT_UNION(x) cout << "[" << type_for_alignment(alignof(x)) << "; " << sizeof(x) / alignof(x) << "]" << endl;
```

Comment (austinwagner) [reply src/shared/wmistr.rs]: That's handy! I think it makes sense to add this block of code and the new `UNION2` style to `CONTRIBUTING.md`. The rules regarding function definition styles and EOF newlines should probably also be there.

Comment (Susurrus) [reply src/shared/wmistr.rs]: @austinwagner `UNION2` is already documented. I just updated #456 about the other things discussed her excluding this example code. I'll add this alignment example to `UNION2` however.

Comment (Susurrus) [reply src/shared/wmistr.rs]: Sorry, you're right that the dual-alignment `UNION2` isn't documented. I'm doing that now.

Comment (Susurrus) [reply src/shared/wmistr.rs]: @austinwagner Does my latest commit for #456 make things clear?

File: src/shared/sddl.rs (hunk 3 above)
Comment (austinwagner) [reply src/shared/sddl.rs]: Does Github ignore EOF newlines? It was missing before, but it's definitely there now. Checked the file directly downloaded from Github in a hex editor to make sure I wasn't going crazy.

**Comment by Susurrus** [comment]
@austinwagner This needs to be updated with the recent merge of #454.

**Comment by Susurrus** [comment]
@austinwagner Just wanted to ping you on this as #455 was just merged and this still needs a rebase.

**Comment by Susurrus** [comment]
@austinwagner This needs another rebase.

**Comment by austinwagner** [comment]
Rebased

**Comment by Susurrus** [comment]
@retep998 Ping!

**Comment by Susurrus** [comment]
@austinwagner Needs a rebase!

**Comment by austinwagner** [comment]
Yikes. Bad merge conflict in winnt.rs. I'll have to spend some time this weekend carefully going through it.

**Comment by Susurrus** [comment]
@austinwagner This needs a rebase. Note that some timezoneapi work was done in #514 which was merged which may complicate this rebase.

**Comment by Susurrus** [comment]
@retep998 This should also be ready for a merge methinks.

**Comment by austinwagner** [comment]
An end to all the rebasing does sound nice ☺

-- Austin Wagner

**Comment by Susurrus** [comment]
@austinwagner #510 was just merged, which finished off all-but-one of the functions left to migrate. I'm not certain if you had any additions here or these were purely migrations, so this may need a rebase or just to be closed depending.

**Comment by retep998** [comment]
This PR is a huge pile of additions, so this PR is still needed.

**Comment by austinwagner** [comment]
Rebaste 🦃

File: src/um/winnt.rs [hunk 5]
```diff
@@ -1948,171 +1970,69 @@ pub const VALID_INHERIT_FLAGS: BYTE = 0x1F;
 pub const SUCCESSFUL_ACCESS_ACE_FLAG: BYTE = 0x40;
 pub const FAILED_ACCESS_ACE_FLAG: BYTE = 0x80;
 pub const TRUST_PROTECTED_FILTER_ACE_FLAG: BYTE = 0x40;
-STRUCT!{struct ACCESS_ALLOWED_ACE {
+STRUCT!{struct __BASE_ACE {
```
Comment (retep998) [inline src/um/winnt.rs]: Why did you replace the contents of all these structures with `__BASE_ACE`?

File: src/um/winnt.rs [hunk 6]
```diff
@@ -2470,10 +2399,21 @@ STRUCT!{struct TOKEN_MANDATORY_LABEL {
 }}
 pub type PTOKEN_MANDATORY_LABEL = *mut TOKEN_MANDATORY_LABEL;
 pub const TOKEN_MANDATORY_POLICY_OFF: DWORD = 0x0;
-pub const TOKEN_MANDATORY_POLICY_NO_WRITE_UP: DWORD = 0x1;
-pub const TOKEN_MANDATORY_POLICY_NEW_PROCESS_MIN: DWORD = 0x2;
-pub const TOKEN_MANDATORY_POLICY_VALID_MASK: DWORD = TOKEN_MANDATORY_POLICY_NO_WRITE_UP
-    | TOKEN_MANDATORY_POLICY_NEW_PROCESS_MIN;
+pub const TOKEN_MANDATORY_POLICY_WRITE_UP: DWORD = 0x1;
+pub const TOKEN_MANDATORY_POLICY_PROCESS_MIN: DWORD = 0x2;
```
Comment (retep998) [inline src/um/winnt.rs]: Why did you change the names of these constants? The names were correct before.

File: tests/headers.rs [hunk 7]
```diff
@@ -67,7 +67,8 @@ fn check_file_deps<P: AsRef<Path>>(
         if let Some(entries) = files_deps.get_mut(&filename) {
             let file_content = read_file(r_p);
             for line in file_content.lines() {
-                if !line.starts_with("use ") && !line.starts_with("pub use ") {
+                if (!line.starts_with("use ") && !line.starts_with("pub use "))
+                    || line.starts_with("use core::") {
```
Comment (retep998) [inline tests/headers.rs]: Instead of modifying the test, just `use core::whatever;` inside the function that needs it.

File: src/um/winnt.rs [hunk 8]
```diff
@@ -2154,15 +2075,15 @@ STRUCT!{struct SECURITY_DESCRIPTOR_RELATIVE {
     Sacl: DWORD,
     Dacl: DWORD,
 }}
-pub type PISECURITY_DESCRIPTOR_RELATIVE = *mut SECURITY_DESCRIPTOR_RELATIVE;
+pub type PSECURITY_DESCRIPTOR_RELATIVE = *mut SECURITY_DESCRIPTOR_RELATIVE;
```
Comment (retep998) [inline src/um/winnt.rs]: `PISECURITY_DESCRIPTOR_RELATIVE` was correct.

File: src/um/winnt.rs [hunk 9]
```diff
@@ -2123,27 +2043,28 @@ STRUCT!{struct ACL_SIZE_INFORMATION {
...
-pub const SECURITY_DESCRIPTOR_MIN_LENGTH: usize = 40;
-#[cfg(target_arch = "x86")]
-pub const SECURITY_DESCRIPTOR_MIN_LENGTH: usize = 20;
+pub const SECURITY_DESCRIPTOR_REVISION: BYTE = 1;
+pub const SECURITY_DESCRIPTOR_REVISION1: BYTE = 1;
+#[cfg(target_arch="x86")]
+const __SIZE_OF_SECURITY_DESCRIPTOR: SIZE_T = 20;
+#[cfg(target_arch="x86_64")]
+const __SIZE_OF_SECURITY_DESCRIPTOR: SIZE_T = 40;
+pub const SECURITY_DESCRIPTOR_MIN_LENGTH: SIZE_T = __SIZE_OF_SECURITY_DESCRIPTOR;
```
Comment (retep998) [inline src/um/winnt.rs]: Don't add the fake `__SIZE_OF` constants. Just hardcode the size into `SECURITY_DESCRIPTOR_MIN_LENGTH` with a `// FIXME` to fix it later when `sizeof` is a const function.

File: src/um/winnt.rs [hunk 10]
```diff
@@ -2123,27 +2043,28 @@ STRUCT!{struct ACL_SIZE_INFORMATION {
...
-pub const SE_SACL_PROTECTED: SECURITY_DESCRIPTOR_CONTROL = 0x2000;
+pub const SE_GROUP_DEFAULTED: SECURITY_DESCRIPTOR_CONTROL = 0x0002;
+pub const SE_OWNER_DEFAULTED: SECURITY_DESCRIPTOR_CONTROL = 0x0001;
 pub const SE_RM_CONTROL_VALID: SECURITY_DESCRIPTOR_CONTROL = 0x4000;
+pub const SE_SACL_AUTO_INHERIT_REQ: SECURITY_DESCRIPTOR_CONTROL = 0x0200;
+pub const SE_SACL_AUTO_INHERITED: SECURITY_DESCRIPTOR_CONTROL = 0x0800;
+pub const SE_SACL_DEFAULTED: SECURITY_DESCRIPTOR_CONTROL = 0x0008;
+pub const SE_SACL_PRESENT: SECURITY_DESCRIPTOR_CONTROL = 0x0010;
+pub const SE_SACL_PROTECTED: SECURITY_DESCRIPTOR_CONTROL = 0x2000;
 pub const SE_SELF_RELATIVE: SECURITY_DESCRIPTOR_CONTROL = 0x8000;
```
Comment (retep998) [inline src/um/winnt.rs]: The old order of these constants was correct. Why did you reorder them to a different order than what is in the header?

File: src/um/winnt.rs [hunk 11]
```diff
@@ -1533,23 +1555,23 @@ STRUCT!{struct SID_AND_ATTRIBUTES {
...
     Hash: [SID_HASH_ENTRY; SID_HASH_SIZE],
 }}
 pub type PSID_AND_ATTRIBUTES_HASH = *mut SID_AND_ATTRIBUTES_HASH;
-pub const SECURITY_NULL_SID_AUTHORITY: [BYTE; 6] = [0, 0, 0, 0, 0, 0];
-pub const SECURITY_WORLD_SID_AUTHORITY: [BYTE; 6] = [0, 0, 0, 0, 0, 1];
-pub const SECURITY_LOCAL_SID_AUTHORITY: [BYTE; 6] = [0, 0, 0, 0, 0, 2];
-pub const SECURITY_CREATOR_SID_AUTHORITY: [BYTE; 6] = [0, 0, 0, 0, 0, 3];
-pub const SECURITY_NON_UNIQUE_AUTHORITY: [BYTE; 6] = [0, 0, 0, 0, 0, 4];
-pub const SECURITY_RESOURCE_MANAGER_AUTHORITY: [BYTE; 6] = [0, 0, 0, 0, 0, 9];
+pub const SECURITY_NULL_SID_AUTHORITY: SID = MAKE_SID!(0);
```
Comment (retep998) [inline src/um/winnt.rs]: Shouldn't these be `SID_IDENTIFIER_AUTHORITY` and not `SID`?

File: src/um/wincrypt.rs [hunk 12]
```diff
@@ -725,7 +725,6 @@ extern "system" {
     pub fn CryptHashSessionKey(
         hHash: HCRYPTHASH,
         hKey: HCRYPTKEY,
-        dwFlags: DWORD,
```
Comment (retep998) [inline src/um/wincrypt.rs]: Why did you get rid of this parameter??

**Comment by retep998** [comment]
PR currently failing CI.
